runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,417 @@
+        client:supports_method(ms.textDocument_foldingRange, bufnr)
+        and (
+          args.data.method == ms.textDocument_didChange
+          or args.data.method == ms.textDocument_didOpen
+        )
+      then
+        request(bufnr, client)
+      end
+    end,
+  })
+
+  for _, client in
+    ipairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+  do
+    request(bufnr, client)
+  end
+
+  return bufstates[bufnr]
+end
+
+-- The "garbage collector" for `bufstate` and its associated event hooks.
+-- See the comment above `augroup_setup` function for more details.
+--
+-- It will be fine without this, as the bufstate and event hooks will also
+-- be cleared when the buffer is unloaded.
+-- api.nvim_create_autocmd('OptionSet', {
+--   group = augroup_setup,
+--   pattern = 'foldexpr',
+--   callback = function()
+--     -- Sould find a more reliable way to detect changes in `foldexpr`.
+--     if vim.v.option_old == 'v:lua.vim.lsp.foldexpr()' then
+--       for bufnr in pairs(bufstates) do
+--         if
+--           ---@param winid integer
+--           vim.iter(vim.fn.win_findbuf(bufnr)):all(function(winid)
+--             return vim.wo[winid].foldexpr ~= 'v:lua.vim.lsp.foldexpr()'
+--           end)
+--         then
+--           bufstates[bufnr] = nil
+--           api.nvim_clear_autocmds({ buffer = bufnr, group = augroup_setup })
+--         end
+--       end
+--     end
+--   end,
+-- })
+
+---@param kind lsp.FoldingRangeKind
+---@param winid integer
+local function foldclose(kind, winid)
+  vim._with({ win = winid }, function()
+    local bufnr = api.nvim_win_get_buf(winid)
+    local row_kinds = bufstates[bufnr].row_kinds
+    -- Reverse traverse to ensure that the smallest ranges are closed first.
+    for row = api.nvim_buf_line_count(bufnr) - 1, 0, -1 do
+      local kinds = row_kinds[row]
+      if kinds and kinds[kind] then
+        vim.cmd(row + 1 .. 'foldclose')
+      end
+    end
+  end)
+end
+
+--- Close all {kind} of folds in the the window with {winid}.
+---
+--- To automatically fold imports when opening a file, you can use an autocmd:
+---
+--- ```lua
+--- vim.api.nvim_create_autocmd('LspNotify', {
+---   callback = function(args)
+---     if args.data.method == 'textDocument/didOpen' then
+---       vim.lsp.folding_range.foldclose('imports', vim.fn.bufwinid(args.buf))
+---     end
+---   end,
+--- })
+--- ```
+---
+---@param kind lsp.FoldingRangeKind "comment" | "imports" | "region"
+---@param winid? integer Defaults to the current window.
+function M.foldclose(kind, winid)
+  vim.validate('kind', kind, 'string')
+  vim.validate('winid', winid, 'number', true)
+
+  winid = winid or api.nvim_get_current_win()
+  local bufnr = api.nvim_win_get_buf(winid)
+  local bufstate = bufstates[bufnr]
+  if not bufstate then
+    return
+  end
+  ---@cast bufstate -nil
+
+  -- Schedule `foldclose()` if the buffer is not up-to-date.
+  if bufstate.version ~= util.buf_versions[bufnr] then
```

Comment:
```suggestion
  if bufstate.version == util.buf_versions[bufnr] then
    foldclose(kind, winid)
    return
  end
```

To reduce code nesting.